### PR TITLE
add notes and questions for n26347

### PR DESCRIPTION
**Pull request by ishaanam**

### CODE REVIEW

File: _posts/2023-01-25-#26347.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+- [WalletRescanReserver](https://github.com/bitcoin/bitcoin/blob/e9262ea32a6e1d364fb7974844fadc36f931f8c6/src/wallet/wallet.h#L951) is used to manage rescans in a wallet.
+
+- Wallets can optionally be encrypted using a passphrase. Only the
+  private key material of the wallet is encrypted using a user-provided
+  passphrase.
+
+- The wallet can be unlocked by calling [walletpassphrase](https://github.com/bitcoin/bitcoin/blob/306ccd4927a2efe325c8d84be1bdb79edeb29b04/src/wallet/rpc/encrypt.cpp#L11),
+  which then stores the passphrase in memory for a specified amount of
+  time.
```
Comment (glozow) [inline _posts/2023-01-25-#26347.md:38]: Light suggestion
```suggestion
  time before the wallet [is locked again](https://github.com/bitcoin/bitcoin/blob/306ccd4927a2efe325c8d84be1bdb79edeb29b04/src/wallet/rpc/encrypt.cpp#L91-L99).
```

File: _posts/2023-01-25-#26347.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
+---
+layout: pr
+date: 2023-01-25
+title: "wallet: ensure the wallet is unlocked when needed for rescanning"
+pr: 26347
+authors: [ishaanam]
+components: ["wallet"]
+host: ishaanam
+status: upcoming
+commit:
```
Comment (glozow) [inline _posts/2023-01-25-#26347.md:10]: Just pushed a tag: https://github.com/bitcoin-core-review-club/bitcoin/commits/pr26347
If you decide to add more links to code/commits in the PR, I recommend using links from that repo. (This is so if the PR branch changes after the review club meeting, people reading the notes can still see what bits of code it's talking about).
```suggestion
commit: b2e2829
```

**Review by glozow — COMMENTED** [review]
Thank you for these excellent notes and questions @ishaanam! A few small suggestions if you have time for them.

File: _posts/2023-01-25-#26347.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  in the current keypool is found to be used, all of the keys up to that
+  key are considered used and the keypool is then topped up.
+
+- [WalletRescanReserver](https://github.com/bitcoin/bitcoin/blob/e9262ea32a6e1d364fb7974844fadc36f931f8c6/src/wallet/wallet.h#L951) is used to manage rescans in a wallet.
+
+- Wallets can optionally be encrypted using a passphrase. Only the
+  private key material of the wallet is encrypted using a user-provided
+  passphrase.
+
+- The wallet can be unlocked by calling [walletpassphrase](https://github.com/bitcoin/bitcoin/blob/306ccd4927a2efe325c8d84be1bdb79edeb29b04/src/wallet/rpc/encrypt.cpp#L11),
```
[the remark this answers is not in the document]
Comment (stickies-v) [reply _posts/2023-01-25-#26347.md:36]: For a more readable alternative, you can also link to the RPC docs instead: https://bitcoincore.org/en/doc/24.0.0/rpc/wallet/walletpassphrase/ (personal preference, so no strong view)

File: _posts/2023-01-25-#26347.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  which then stores the passphrase in memory for a specified amount of
+  time.
+
+- Certain RPCs which make use of information relating to private keys
```
Comment (stickies-v) [inline _posts/2023-01-25-#26347.md:40]: Could be useful to provide references to 1/2 of those RPCs

File: _posts/2023-01-25-#26347.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+## Questions
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
+
+1. What bug is this PR fixing?
+
+1. Why can wallets that use unhardened derivation top up the keypool
+without being unlocked? Would this problem occur when rescanning a
+descriptor wallet with descriptors generated by Bitcoin Core?
+
+1. Does the [test](https://github.com/bitcoin/bitcoin/pull/26347/commits/b2e2829014c5c9a3dc02eab7fcd77dd907228c78)
```
Comment (stickies-v) [inline _posts/2023-01-25-#26347.md:63]: As per glozow's other suggestion:
```suggestion
1. Does the [test](https://github.com/bitcoin-core-review-club/bitcoin/commit/b2e2829014c5c9a3dc02eab7fcd77dd907228c78)
```

**Review by stickies-v — COMMENTED** [review]
Awesome to see you host a review club @ishaanam, thank you! Very nice preparation and questions. I agree with all of @glozow's comments that pretty much cover it all, just added a few more minor suggestions.

File: _posts/2023-01-25-#26347.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+of this PR was removed and a user tried to run `walletlock` while
+performing a rescan that required the passphrase?
+
+1. Why was another mutex added instead of just using `cs_wallet`?
+
+1. Why is `RecursiveMutex` used here as opposed to `Mutex`?
+
+1. Why does `m_relock_mutex` get locked twice in `walletlock`?
+
+1. Can you think of any alternative approaches that could fix this bug?
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply _posts/2023-01-25-#26347.md]: Yes, this order does make more sense.

File: _posts/2023-01-25-#26347.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ishaanam) [reply _posts/2023-01-25-#26347.md]: I have linked to the RPC docs now.

**Comment by ishaanam** [comment]
Thanks for the suggestions @glozow and @stickies-v! I have now addressed them.

File: _posts/2023-01-25-#26347.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+- An encrypted wallet is [locked](https://github.com/bitcoin/bitcoin/blob/306ccd4927a2efe325c8d84be1bdb79edeb29b04/src/wallet/wallet.cpp#L3358) by default;
+  it can be unlocked by calling [`walletpassphrase`](https://bitcoincore.org/en/doc/24.0.0/rpc/wallet/walletpassphrase/),
+  which then stores the passphrase in memory for a specified amount of
+  time before the wallet [is locked again](https://github.com/bitcoin/bitcoin/blob/306ccd4927a2efe325c8d84be1bdb79edeb29b04/src/wallet/rpc/encrypt.cpp#L91-L99)
+
+- Certain RPCs which make use of information relating to private keys, such as
+  [`sendtoaddress`](https://github.com/bitcoin/bitcoin/blob/39363a4b945114f5e4718f75098f3036e8fe6a1d/src/wallet/rpc/spend.cpp#L212) and [`signrawtransactionwithwallet`](https://github.com/bitcoin/bitcoin/blob/39363a4b945114f5e4718f75098f3036e8fe6a1d/src/wallet/rpc/spend.cpp#L840)
+  require the wallet to be unlocked in order to work correctly. Typically if an RPC
+  requires the passphrase, it throws an error if it is not in memory.
```
Comment (stickies-v) [inline _posts/2023-01-25-#26347.md:45]: nit: I'd avoid referring to the same thing with different terminologies
```suggestion
  require the wallet to be unlocked in order to work correctly. If the wallet is
  not unlocked, the RPC typically throws an error.
```

**Review by stickies-v — COMMENTED** [review]
Thanks - all comments seem addressed! Just two further follow up nits on the changes.

File: _posts/2023-01-25-#26347.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  of keys, descriptors, etc. A rescan can also be invoked by calling
+  [`rescanblockchain`](https://github.com/bitcoin/bitcoin/blob/306ccd4927a2efe325c8d84be1bdb79edeb29b04/src/wallet/rpc/transactions.cpp#L844).
+
+- Wallet rescans were previously discussed in review clubs
+  [#15845](/15845) and [#25957](/25957).
+
+- When a wallet is rescanning, it needs to be able to top-up its keypool.
+
+- A wallet's keypool is its reserve of unused keys. An unused key is a key that
+  hasn't been sent funds yet. [`MarkUnusedAddresses`](https://github.com/bitcoin/bitcoin/blob/39363a4b945114f5e4718f75098f3036e8fe6a1d/src/wallet/scriptpubkeyman.cpp#L2182)
```
[the remark this answers is not in the document]
Comment (ishaanam) [reply _posts/2023-01-25-#26347.md]: I have now clarified this point.

**Review by stickies-v — APPROVED** [review]
ACK, thanks @ishaanam for the prep and have fun with the hosting! Don't hesitate to reach out if you have any questions, more than happy to help.

**Pull request merged**